supervisor/iptablesctrl/uidpam/update_portsets.go

```diff
@@ -0,0 +1,260 @@
+		p.UIDToPorts.Add(uid, portList)
+	}
+
+	return ok, nil
+}
+
+// AddToUIDToPortSetCache : This adds/updates UIDToPortSet cache. This cache
+// maps UID to the portset associated with its PU. This is called during
+// creation of PU/on reception of application SYN-ACK packet.
+func (p *UIDCacheInstance) AddToUIDToPortSetCache(username string, portset string) (err error) {
+	uid, err := p.getUserID(username)
+	if err != nil {
+		return err
+	}
+
+	p.UIDToPortSet.AddOrUpdate(uid, portset)
+	p.UIDSet[uid] = true
+	return nil
+
+}
+
+// GetFromUIDToPortSetCache : This returns the portset associated with UID.
+func (p *UIDCacheInstance) GetFromUIDToPortSetCache(username string) (port string, err error) {
+	uid, err := p.getUserID(username)
+	if err != nil {
+		return "", err
+	}
+	portSetName, err := p.UIDToPortSet.Get(uid)
+	if err != nil {
+		return "", err
+	}
+	return portSetName.(string), nil
+}
+
+// RemoveFromUIDToPortSetCache : This deletes UID from UIDToPortSet cache. This is called
+// during removal of PU.
+func (p *UIDCacheInstance) RemoveFromUIDToPortSetCache(username string) (err error) {
+	uid, err := p.getUserID(username)
+	if err != nil {
+		return err
+	}
+	delete(p.UIDSet, uid)
+	return p.UIDToPortSet.Remove(uid)
+}
+
+// RemoveFromUIDToPortsCache : Is this ever required ? Go routine updates the Port
+func (p *UIDCacheInstance) RemoveFromUIDToPortsCache(uid string) (err error) {
+	return nil
+}
+
+// UpdatePortSet : This API programs the ipset portset with port. The
+// portset name is derived from UIDToPortSet cache.
+func (p *UIDCacheInstance) UpdatePortSet(username string, port string) (err error) {
+
+	puPortSetName, err := p.GetFromUIDToPortSetCache(username)
+	if err != nil {
+		return fmt.Errorf("Unable to get portset from uid")
+	}
+
+	ips := ipset.IPSet{
+		Name: puPortSetName,
+	}
+
+	if _, err := strconv.Atoi(port); err != nil {
+		return fmt.Errorf("Not a valid Port")
+	}
+
+	if adderr := ips.Add(port, portEntryTimeout); adderr != nil {
+		return fmt.Errorf("Unable to add port to set")
+	}
+	return nil
+}
+
+// InitPortSetTask This go routine periodically scans (1s)
+// /proc/net/tcp file for listening ports and programs
+// the portsets. This worker thread is setup during datapath
+// initilisation.
+func InitPortSetTask(p *UIDCacheInstance) {
+	go startPortSetTask(p)
+}
+
+func startPortSetTask(p *UIDCacheInstance) {
+	t := time.NewTicker(portSetUpdateIntervalMilliseconds * time.Millisecond)
+	for range t.C {
+		// Update PortSet periodically.
+		p.updateIPPortSets()
+	}
+}
+
+func (p *UIDCacheInstance) updateIPPortSets() {
+	file, err := os.Open(procNetTCPFile)
+	if err != nil {
+		zap.L().Warn("Failed To open /proc/net/tcp file", zap.Error(err))
+		return
+	}
+	defer func() {
+		err := file.Close()
+		if err != nil {
+			zap.L().Warn("Failed To close /proc/net/tcp file", zap.Error(err))
+		}
+	}()
+
+	localCache := cache.NewCache("localCache")
+	scanner := bufio.NewScanner(file)
+	lineCnt := 0
+	for scanner.Scan() {
```

Comment:
This should be like a regex match where we get a string which is port list which is then split using strings.Split as opposed to hardcoded parsing logic.